### PR TITLE
Implement low stock and metadata config file for devices

### PR DESCRIPTION
**Pull request by brentstineman**
closes item #2

### CODE REVIEW

File: src/Devices/Base/BaseDevice.cs [hunk 1]
```diff
@@ -4,49 +4,39 @@
...
-        private string connectionString;
-        protected dynamic deviceSettings; 
+        protected IEventScheduler eventScheduler;
+        protected IDeviceClient deviceClient;
 
         protected string deviceId;
         protected string deviceType;
 
-        public BaseDevice(string deviceConfigFile, string connectionString)
+        public BaseDevice(IDeviceConfig deviceConfig, IDeviceClient client, IEventScheduler eventScheduler)
```
Comment (Sarah-Aly) [inline src/Devices/Base/BaseDevice.cs:22]: Params need to be validated.

File: src/Devices/Base/ConfigurationHandler.cs [hunk 2]
```diff
@@ -20,11 +21,17 @@ public static string getConfig(string section, string key)
...
-            // configurationSection.Value => C:\\temp\\logs\\output.txt
+
             IConfigurationSection configurationSection = configuration.GetSection(section).GetSection(key);
             return configurationSection.Value;
 
         }
+
+        public static string GetDeviceRuntimeSettings(string configurationSetting)
+        {
+            // TODO: put some error handling around the file access
```
Comment (Sarah-Aly) [inline src/Devices/Base/ConfigurationHandler.cs]: Please, remove the TODO, fix it, or create a work item for fixing it & link it here.

File: src/Devices/Base/TimedSimulatedEvent.cs [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
 
 namespace Transportation.Demo.Devices.Base
 {
-    public class SimulatedEvent
+    public class TimedSimulatedEvent : ISimulatedEventWithSetter
     {
         private Timer eventTimer;
         private double interval;
         private double jitter; 
         // return true is ready to continue, otherwise return false
```
Comment (Sarah-Aly) [inline src/Devices/Base/TimedSimulatedEvent.cs]: Please, remove this comment.

File: src/Devices/Kiosk/KioskDevice.cs [hunk 4]
```diff
@@ -74,7 +84,7 @@ private Task<MethodResponse> ReceivePurchaseTicketResponse(MethodRequest methodR
             string result = "{\"result\":\"Executed direct method: " + methodRequest.Name + "\"}";
 
             // restart timer
-            purchaseEvent.Start();
+            this.eventScheduler.StartAll(); // TODO: need more elegant solution
```
Comment (Sarah-Aly) [inline src/Devices/Kiosk/KioskDevice.cs]: Please, remove the TODO, fix it, or create a work item for fixing it & link it here.

File: src/Devices/Base/BaseDevice.cs (hunk 1 above)
Comment (brentstineman) [reply src/Devices/Base/BaseDevice.cs:22]: added validation

File: src/Devices/Base/ConfigurationHandler.cs (hunk 2 above)
Comment (brentstineman) [reply src/Devices/Base/ConfigurationHandler.cs]: Added error handling and validation of device config file

File: src/Devices/Base/TimedSimulatedEvent.cs (hunk 3 above)
Comment (brentstineman) [reply src/Devices/Base/TimedSimulatedEvent.cs]: removed,

File: src/Devices/Kiosk/KioskDevice.cs (hunk 4 above)
Comment (brentstineman) [reply src/Devices/Kiosk/KioskDevice.cs]: removed and used a different approach. May revisit use of Observer pattern and reactive extensions at a later date if time allows.

File: src/Devices/Kiosk/KioskDevice.cs [hunk 5]
```diff
@@ -84,7 +86,7 @@ private Task<MethodResponse> ReceivePurchaseTicketResponse(MethodRequest methodR
             string result = "{\"result\":\"Executed direct method: " + methodRequest.Name + "\"}";
 
             // restart timer
-            this.eventScheduler.StartAll(); // TODO: need more elegant solution
+            this.purchaseticketEvent.Start(); // TODO: need more elegant solution
```
Comment (Sarah-Aly) [inline src/Devices/Kiosk/KioskDevice.cs]: This removes the event scheduler abstraction. This will again interfere in the testing. I wonder why the change is needed to begin with, given that the Kiosk device owns the scheduler now, and it knows that it only has purchase ticket event.

File: src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs [hunk 6]
```diff
@@ -102,6 +102,37 @@ public void TestKioskPurchaseTicket()
...
+                DeviceId = deviceconfig.DeviceId,
+                DeviceType = deviceconfig.DeviceType,
+                MessageType = "Purchase",
+                IsApproved = false
+            };
+
+            // create our test device
+            KioskDevice device = new KioskDevice(deviceconfig, fakeDeviceClient, fakeScheduler);
+
+            TestContext.WriteLine("/n>> Purchasing tickets, shouldn't throw event");
```
Comment (Sarah-Aly) [inline src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs]: did you mean "\n" ? If that's the case, please make an empty WriteLine instead

Comment (brentstineman) [reply src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs]: yes, its there's to make the test log more readable. Are you sure we need an entirely new line of code just for readability?

File: src/Devices/Kiosk/KioskDevice.cs (hunk 5 above)
Comment (brentstineman) [reply src/Devices/Kiosk/KioskDevice.cs]: the tests actually still function without error. The primary reason is that if the class is enhanced in the future with additional timed events (like a simulated device failure), we don't restart everything. We really only want to restart the one device. 

The observer pattern would have been idea for this. But we need to get this in ASAP as we're already weeks behind schedule. So I'm opting or less elegant solution.

Comment (Sarah-Aly) [reply src/Devices/Kiosk/KioskDevice.cs]: When the future is there, we can add the capability on the scheduler to manage a specific event rather than all. For now, this does interfere with the testing (just didn't have the chance yet -- and in very high CPU test runs, it very well might have that chance), and it also is not verifiable, in the sense that our test cannot verify now that starts been called by the callback. Please change it back for now, and if you have time add validation in the test that start all on the scheduler has been called

File: src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs (hunk 6 above)
Comment (Sarah-Aly) [reply src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs]: I highly prefer splitting them for the following reasons:
* It's more readable in the Code, more explicit, and matches the output better.
* It's more standard, in Windows we'll need the carriage return too.
Anyway it's up to you, but at the very least I believe you need to fix the /n to \n.

File: src/Devices/Kiosk/KioskDevice.cs (hunk 5 above)
Comment (brentstineman) [reply src/Devices/Kiosk/KioskDevice.cs]: "start all" has been called will require some effort as the fakeScheduler doesn't track a start/stop value for the timer. I'll get both in shortly.

File: src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs (hunk 6 above)
Comment (brentstineman) [reply src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs]: kk

Comment (brentstineman) [reply src/TransitFunctionAppTest/ModelsTests/KioskDeviceTest.cs]: changed

File: src/Devices/Kiosk/KioskDevice.cs (hunk 5 above)
Comment (brentstineman) [reply src/Devices/Kiosk/KioskDevice.cs]: updated. FakeSimulatedEvent will now track and report a simulated running state.

**Comment by Sarah-Aly** [comment]
I'm approving the PR but I highly suggest to have the fake scheduler extend the real one and just override the add method :)

**Comment by brentstineman** [comment]
I tried that and had some issues. I'll try to sort that out while working on the next work item (#6). I'll update that work item to call out that change as well.
